libs/langchain/langchain/document_loaders/parsers/pdf_docai.py

```diff
@@ -0,0 +1,217 @@
+"""Module contains a PDF parser based on DocAI from Google Cloud.
+
+You need to install two libraries to use this parser:
+pip install google-cloud-documentai
+pip install google-cloud-documentai-toolbox
+"""
+import time
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Iterator, List, Optional
+
+from google.api_core.operation import Operation
+
+from langchain.docstore.document import Document
+from langchain.document_loaders.base import BaseBlobParser
+from langchain.document_loaders.blob_loaders import Blob
+
+if TYPE_CHECKING:
+    from google.cloud.documentai import DocumentProcessorServiceClient
+
+
+@dataclass
+class DocAIParsingResults:
+    """A dataclass to store DocAI parsing results."""
+
+    source_path: str
+    parsed_path: str
+
+
+class DocAIPdfLoader(BaseBlobParser):
+    def __init__(
+        self,
+        project: str,
+        location: str = "us",
+        processor_name: Optional[str] = None,
+        gcs_output_path: Optional[str] = None,
+    ):
+        from google.api_core.client_options import ClientOptions
+
+        self._options = ClientOptions(
+            api_endpoint=f"{location}-documentai.googleapis.com"
+        )
+        self._processor_name = processor_name
+        self._name = self.client.common_location_path(project=project, location="us")
+        self._gcs_output_path = gcs_output_path
+
+    def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+        """Parses a blob lazily.
+
+        This is a long-running operations! A recommended way is to batch
+            documents together and use `batch_parse` method.
+        """
+        yield from self.batch_parse(
+            [blob], self._gcs_output_path  # type: ignore[arg-type]
+        )
+
+    def batch_parse(
+        self, blobs: List[Blob], gcs_output_path: str
+    ) -> Iterator[Document]:
+        """Parses a list of blobs lazily.
+
+        Args:
+            blobs: a list of blobs to parse
+            gcs_output_path: a path on GCS to store parsing results
+        This is a long-running operations! A recommended way is to decouple
+            parsing from creating Langchain Documents:
+            >>> operations = parser.docai_parse(blobs, gcs_path)
+            >>> parser.is_running(operations)
+            You can get operations names and save them:
+            >>> names = [op.name for op in operations]
+            And when all operations are finished, you can use their results:
+            >>> operations = parser.operations_from_names(operation_names)
+            >>> results = parser.get_results(operations)
+            >>> docs = parser.batch_parse_from_results(results)
+        """
+        operations = self.docai_parse(blobs, gcs_output_path=gcs_output_path)
+        is_running = self.is_running(operations)
```

Comment:
Save an extra request: 

```
while True:
   if not self.is_running(operations):
     break
```      

It may make sense to add a timeout parameter here

Reply:
Done